### DISCUSSION TITLE
instead of `append-at`, what if we have `$children`

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#380] [Ideas & RFCs]
Status: Being implemented as part of [end syntax](https://github.com/FifthTry/ftd/discussions/381).

Currently a variable `append-at` controls where the children are placed. If `append-at` is not specified, the last open container gets the external children.

What if instead of this we had explicit `$children` attribute, and component authors have to explicitly pass this variable to wherever they want it to show up at.

Which explicitly tells this component accepts children by looking at $children (we can make it explicit argument, or implicit, if you used $children, you must accept $children. leaning towards explicit).

**Comment by amitu** [comment]
An example, say `foo` is defined like this:

```ftd
-- ftd.row foo:
```

When we pass children to `foo`, they auto magically go in the `ftd.row`.

```ftd
-- foo:
-- ftd.text: hello
```

Generates:

```
ftd.row
  - ftd.text
```

The rule was "whatever is the **current container** at the end of definition of `foo` gets the children". 

In proposed syntax we will have to define `foo` as:

```ftd
-- ftd.row foo:
children: $children
```

or 

```ftd
-- ftd.row foo:
ftd.ui list children:
children: $children
```

# `append-at` example

```ftd
-- ftd.column foo:
append-at: x

-- ftd.row:
id: x

-- container: ftd.main
-- ftd.row:
```

Without `append-at`, children of `foo` would be passed to the second row, but with `append-at` they will be passed to first row.

In the new syntax:

```ftd
-- ftd.column foo:
ftd.ui list children:

-- ftd.row:
children: $children

-- container: ftd.main
-- ftd.row:
```

**Comment by Arpita-Jaiswal** [comment]
`append-at` vs `children`

Currently, when we look at component definitions `append-at` and `open`, it is clear that they accept children. But with proposed design, we have to scan through whole definition to know if it accepts children.

Two counterpoints:

1. We are explicitly taking children as an argument 
2. We always don't specify `open` and `append-at`.

**Comment by Arpita-Jaiswal** [comment]
With proposed design we accept children as an argument in the header. `ftd.ui list children`, this looks like we need to pass a list `children: $x`.

We already have a caption and body which accepts the value from different places.

**Comment by Arpita-Jaiswal** [comment]
Can we allow the user to rebind the `ftd-ui list children` to some variable-name `ftd-ui list x`? 

No, children is a special name like caption and body. Also it's never a good idea to rebind it just adds confusion.
